extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import * as styles from './base-carousel.css';
+import {Scroller} from './scroller';
+import {useStateFromProp} from '../../../src/preact/utils';
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function BaseCarousel(props) {
+  const {style, arrowPrev, arrowNext, children, currentSlide} = props;
+  const {0: curSlide, 1: setCurSlide} = useStateFromProp(currentSlide || 0);
+
+  return (
+    <div style={style}>
+      <Scroller currentSlide={curSlide} setCurrentSlide={setCurSlide}>
+        {children}
+      </Scroller>
+      <Arrow
+        dir={-1}
+        currentSlide={curSlide}
+        setCurrentSlide={setCurSlide}
+        length={children.length}
+        customArrow={arrowPrev}
+      ></Arrow>
+      <Arrow
+        dir={1}
+        currentSlide={curSlide}
+        setCurrentSlide={setCurSlide}
+        length={children.length}
+        customArrow={arrowNext}
+      ></Arrow>
+    </div>
+  );
+}
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Arrow(props) {
+  const {dir, currentSlide, setCurrentSlide, length, customArrow} = props;
+  const button = customArrow ? customArrow : DefaultArrow({dir});
```

Comment:
This is interesting, but also pretty dangerous. Essentially `DefaultArrow` cannot contain any hooks. Maybe it'd be better to name this method something like `renderDefaultArrow()` to make this clear?

Reply:
Done

Reply:
for my learning: why can't DefaultArrow have hooks?